wdl-engine/src/tree.rs

```diff
@@ -0,0 +1,543 @@
+            .0
+            .green
+            .as_node()
+            .expect("should be node")
+            .children()
+            .enumerate()
+        {
+            if let Some((_, prev)) = last {
+                start += usize::from(prev.text_len());
+            }
+
+            last = Some((index, child));
+        }
+
+        match last? {
+            (index, NodeOrToken::Node(n)) => {
+                self.new_child_node(n.to_owned(), index, start).last_token()
+            }
+            (index, NodeOrToken::Token(t)) => {
+                Some(self.new_child_token(t.to_owned(), index, start))
+            }
+        }
+    }
+
+    fn descendants(&self) -> impl Iterator<Item = Self> {
+        self.preorder().filter_map(|event| match event {
+            WalkEvent::Enter(node) => Some(node),
+            WalkEvent::Leave(_) => None,
+        })
+    }
+
+    fn ancestors(&self) -> impl Iterator<Item = Self> {
+        std::iter::successors(Some(self.clone()), SyntaxNode::parent)
+    }
+
+    fn is_rule_excepted(&self, _: &str) -> bool {
+        // For engine evaluation, we except all rules
+        true
+    }
+}
+
+impl fmt::Debug for SyntaxNode {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if f.alternate() {
+            let mut level = 0;
+            for event in self.preorder_with_tokens() {
+                match event {
+                    WalkEvent::Enter(element) => {
+                        for _ in 0..level {
+                            write!(f, "  ")?;
+                        }
+                        match element {
+                            NodeOrToken::Node(node) => writeln!(f, "{:?}", node)?,
+                            NodeOrToken::Token(token) => writeln!(f, "{:?}", token)?,
+                        }
+                        level += 1;
+                    }
+                    WalkEvent::Leave(_) => level -= 1,
+                }
+            }
+            assert_eq!(level, 0);
+            Ok(())
+        } else {
+            write!(f, "{:?}@{}", self.kind(), self.span())
+        }
+    }
+}
+
+impl NewRoot<wdl_ast::SyntaxNode> for SyntaxNode {
+    fn new_root(root: wdl_ast::SyntaxNode) -> Self {
+        Self(Arc::new(ElementData::new_root(root.green().into())))
+    }
+}
+
+impl From<SyntaxNode> for SyntaxElement {
+    fn from(value: SyntaxNode) -> Self {
+        Self::Node(value)
+    }
+}
+
+/// Represents a syntax token that is `Send`+`Sync`.
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct SyntaxToken(Arc<ElementData>);
+
+impl SyntaxToken {
+    /// Gets the next sibling node or token.
+    ///
+    /// Returns `None` if there is no next sibling node or token.
+    pub fn next_sibling_or_token(&self) -> Option<SyntaxElement> {
+        // This should also be a constant-time access rather than having to iterate
+        // We need the offset relative to the start of the parent from the green node to
+        // do that
```

Comment:
```suggestion
        // This should also be a constant-time access rather than having to iterate.
        // We need the offset relative to the start of the parent from the green node to
        // do that.
```